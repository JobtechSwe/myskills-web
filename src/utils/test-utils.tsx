import * as React from 'react'
import ApolloClient from 'apollo-client'
import { ApolloProvider } from 'react-apollo-hooks'
import { InMemoryCache } from 'apollo-cache-inmemory'
import { MockLink, MockedProvider } from 'react-apollo/test-utils'
import { ThemeProvider } from 'emotion-theming'
import { render as rtlRender } from 'react-testing-library'
import { theme } from 'theme'

export const render = (children: React.ReactNode, mocks?: any[]) => {
  const client = new ApolloClient({
    cache: new InMemoryCache({ addTypename: true }),
    link: new MockLink(mocks || []),
  })

  return rtlRender(
    <MockedProvider addTypename={true} mocks={mocks}>
      <ApolloProvider client={client}>
        <ThemeProvider theme={theme}>{children}</ThemeProvider>
      </ApolloProvider>
    </MockedProvider>
  )
}
