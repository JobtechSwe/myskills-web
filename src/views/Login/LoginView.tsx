import React from 'react'
import { OpenInApp } from 'components/ButtonLink'
import styled from '@emotion/styled'
import Flex from '../../components/Flex'
import Grid from '../../components/Grid'
import logotypeIcon from '../../assets/images/logo.svg'
import egendataLogotypeIcon from '../../assets/icons/egendata_logo.svg'
import egendataLogotypeIconWhite from '../../assets/icons/egendata_logo_white.svg'
import { H3, Paragraph } from '../../components/Typography'
import Info from '../../components/Info'
import { Link } from '../../components/Link'
import check from '../../assets/icons/check_blue.svg'

interface LoginViewProps {
  loginUrl: string
}

const Wrapper = styled.div`
  align-items: center;
  background: ${({ theme }) => `radial-gradient(
    647.69px at 6.66% 96.53%,
    ${theme.colors.yourPink} 0%,
    ${theme.colors.seashellPeach} 100%
  )`};
  display: flex;
  justify-content: center;
  padding: 25px;
  height: 100vh;
`

const LogosWrapper = styled.div`
  margin: 50px 0;
  text-align: center;
`

const Image = styled.img`
  margin-top: 1px;
  margin-left: 7px;
`

const InfoBulletsWrapper = styled.div`
  margin: 40px 0;
  padding: 0 30px;
`

const LoginView: React.FC<LoginViewProps> = ({ loginUrl }) => {
  return (
    <>
      <Wrapper>
        <Flex
          alignItems="center"
          flexDirection="column"
          justifyContent="center"
        >
          <LogosWrapper>
            <img alt="logo" src={logotypeIcon} />
            <Paragraph mb={10} mt={10}>
              vill öpnna
            </Paragraph>
            <img alt="egendatalogo" src={egendataLogotypeIcon} />
          </LogosWrapper>
          <OpenInApp url={loginUrl}>
            Logga in med{' '}
            <Image alt="egendatalogo" src={egendataLogotypeIconWhite} />
          </OpenInApp>
          <InfoBulletsWrapper>
            <Grid gridTemplateColumns="20px 1fr">
              <img alt="check" src={check} />
              <div>
                <H3 color="black" fontSize={16} fontWeight="bold" mb="small">
                  Få kontroll över din data
                </H3>
                <Paragraph color="black" mb="large" mt="small">
                  Genom Egendata väljer du själv var din information ska sparas.
                </Paragraph>
              </div>
            </Grid>
            <Grid gridTemplateColumns="20px 1fr">
              <img alt="check" src={check} />
              <div>
                <H3 color="black" fontSize={16} fontWeight="bold" mb="small">
                  Spara tid
                </H3>
                <Paragraph color="black" mb="large" mt="small">
                  Hämta och dela din information istället för att skapa ett nytt
                  CV för varje jobbsökarsajt.
                </Paragraph>
              </div>
            </Grid>
          </InfoBulletsWrapper>
          <Info>
            <Link textDecoration="underline">
              Läs mer om hur Egendata fungerar
            </Link>
          </Info>
        </Flex>
      </Wrapper>
    </>
  )
}

export default LoginView
