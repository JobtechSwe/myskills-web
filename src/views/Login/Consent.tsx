import React from 'react'
import gql from 'graphql-tag'
import { useSubscription } from 'react-apollo-hooks'
import QR from '../../components/QR/QR'
import { navigate } from '@reach/router'
import { setCookie } from '../../utils/helpers'

export const CONSENT_SUBSCRIPTION = gql`
  subscription consentApproved($consentRequestId: String!) {
    consentApproved(consentRequestId: $consentRequestId) {
      accessToken
    }
  }
`

interface IConsentProps {
  consentId: string
}

const Consent: React.FC<IConsentProps> = ({ consentId }) => {
  const { data, error, loading } = useSubscription(CONSENT_SUBSCRIPTION, {
    variables: {
      consentRequestId: consentId,
    },
  })

  const renderConsentStatus = () => {
    if (loading) {
      return <p>Waiting for consent...</p>
    }

    if (error) {
      return <p>That is an error...</p>
    }

    if (data) {
      setCookie('token', data.consentApproved.accessToken)
      navigate('/profile')
      return null
    }
  }

  return (
    <>
      {renderConsentStatus()}
      <QR consentId={consentId} />
      <p id="consentId">{consentId}</p>
    </>
  )
}

export default Consent
