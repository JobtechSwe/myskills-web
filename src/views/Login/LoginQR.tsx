import React from 'react'
import gql from 'graphql-tag'
import { useSubscription } from 'react-apollo-hooks'
import QR from '../../components/QR'
import { navigate } from '@reach/router'
import { setCookie } from '../../utils/helpers'

export const LOGIN_SUBSCRIPTION = gql`
  subscription loginApproved($loginRequestId: String!) {
    loginApproved(loginRequestId: $loginRequestId) {
      accessToken
    }
  }
`

interface ILoginQRProps {
  loginUrl: string
  loginRequestId: string
}

const LoginQR: React.FC<ILoginQRProps> = ({ loginRequestId, loginUrl }) => {
  const { data, error, loading } = useSubscription(LOGIN_SUBSCRIPTION, {
    variables: {
      loginRequestId,
    },
  })

  const renderLoginStatus = () => {
    if (loading) {
      return <p>Waiting for login approval...</p>
    }

    if (error) {
      return <p>That is an error...</p>
    }

    if (data) {
      setCookie('token', data.loginApproved.accessToken)
      navigate('/profile')
      return null
    }
  }

  return (
    <>
      {renderLoginStatus()}
      <QR consentId={loginUrl} />
      <p id="consentId">{loginUrl}</p>
    </>
  )
}

export default LoginQR
