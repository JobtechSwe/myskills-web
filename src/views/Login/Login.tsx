import React from 'react'
import gql from 'graphql-tag'
import { Mutation } from 'react-apollo'
import { RouteComponentProps } from '@reach/router'
import Button from '../../components/Button'
import { Paragraph } from '../../components/Typography'
import LoginView from './LoginView'

export const GET_LOGIN_ID = gql`
  mutation login {
    login {
      url
      sessionId
    }
  }
`

const Login: React.FC<RouteComponentProps> = () => {
  return (
    <>
      {/* TODO(@all):
       *  Replace this with useMutation when support has been added:
       *  https://github.com/trojanowski/react-apollo-hooks/pull/93
       */}
      <Mutation mutation={GET_LOGIN_ID}>
        {(login, { data, error, loading }) => {
          if (loading) {
            return <Paragraph>Loading...</Paragraph>
          }

          if (error) {
            return <Paragraph>Thatâ€™s an error.</Paragraph>
          }

          if (data) {
            return <LoginView loginUrl={data.login.url} />
          }

          return (
            <Button onClick={(_e: any) => login()} variant="primary">
              Login
            </Button>
          )
        }}
      </Mutation>
    </>
  )
}

export default Login
