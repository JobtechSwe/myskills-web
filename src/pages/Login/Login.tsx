import React from 'react'
import gql from 'graphql-tag'
import { Mutation } from 'react-apollo'
import { RouteComponentProps } from '@reach/router'
import styled from '@emotion/styled'
import Consent from './Consent'

const Test = styled.div`
  width: 100vw;
  height: 100vh;
  display: flex;
  flex-flow: column nowrap;
  justify-content: center;
  align-items: center;
  background: ${({ theme }) => theme.colors.lassekongo};
`

export const GET_CONSENT_ID = gql`
  mutation login {
    login {
      id
      expires
    }
  }
`

const Login: React.FC<RouteComponentProps> = props => {
  return (
    <Test>
      {/* TODO(@all):
       *  Replace this with useMutation when support has been added:
       *  https://github.com/trojanowski/react-apollo-hooks/pull/93
       */}
      <Mutation mutation={GET_CONSENT_ID}>
        {(login, { data, error, loading }) => {
          if (loading) {
            return <p>Loading...</p>
          }

          if (error) {
            return <p>Thatâ€™s an error.</p>
          }

          if (data) {
            return <Consent consentId={data.login.id} />
          }

          return <button onClick={_e => login()}>LOGIN</button>
        }}
      </Mutation>
    </Test>
  )
}

export default Login
