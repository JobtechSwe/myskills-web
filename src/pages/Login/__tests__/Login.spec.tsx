import * as React from 'react'
import { MockedProvider } from 'react-apollo/test-utils'
import { render } from 'react-testing-library'
import Login, { GET_CONSENT_ID } from '../Login'

const mocks = [
  {
    request: {
      query: GET_CONSENT_ID,
    },
    result: {
      data: {
        consent: { id: '123456789', expires: '123' },
      },
    },
  },
]

it('renders without errors', async () => {
  const { container } = render(
    <MockedProvider addTypename={false} mocks={mocks}>
      <Login path="/" />
    </MockedProvider>
  )

  // to pass loading state lib or other solution for this?
  await new Promise(res => setTimeout(res, 0))

  expect(container).toMatchSnapshot()
})
