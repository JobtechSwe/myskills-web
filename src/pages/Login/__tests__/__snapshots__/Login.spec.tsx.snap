// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages/Login renders without errors 1`] = `
<div>
  <div
    class="css-1lai7io"
  >
    <button>
      LOGIN
    </button>
  </div>
</div>
`;
